### PR TITLE
Refactor userQueryParams for Outbounds

### LINKED ISSUE #1204: Cannot delete the Outbound shipment
### Step to reproduce

- Setup remote-server, Central server, Web-client
- Login to Web-client
- Create a dummy Outpatient invoice, add a line then delete the line save it
- View the Outpatient list
- Find that Outpatient shipment
- Select it and click `Delete selected lines`  from the option DDL
- It asked to `Select the line / lines`, but is already selected

Screen Shot 2022-04-29 at 1 45 54 PM

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #1204 

Am creating the PR at this stage.. there is more refactoring to do, but I want to reduce the number of files in the PR 😉 
This one is the essence of the change - the next steps are:
- replicate these changes across all usages of `useQueryParams`
- remove the hooks `useFilterBy` `useSortBy` & `usePagination`
- rework existing tests

The core of the problem is that the current hooks aren't sharing state.. so a list of items is retrieved by two separate calls, and each has different query params, though the code is expecting to see the same list of results.

I've lifted the queryParams state out, so it is created in the page context. Almost always, this is within the table context, so have created a query params context which is instantiated by the table context. For the cases where the queryParams are used outside of the table context (Name search is one) then the query params context can be instantiated separately.

Will also replace the `useSharedSort` implementation required by this PR: https://github.com/openmsupply/openmsupply-client/pull/1017 as I expect that won't be needed now.

### CODE REVIEW

File: packages/common/src/hooks/useFilterBy/useFilterBy.tsx [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import { useState } from 'react';
 
-interface FilterByConditionByType {
+export interface FilterByConditionByType {
```
Comment (mark-prins) [inline packages/common/src/hooks/useFilterBy/useFilterBy.tsx:3]: only the types will be used from the existing hooks, so will just lift those out

File: packages/common/src/hooks/useQueryParams/useQueryParams.ts [hunk 2]
```diff
@@ -48,3 +67,140 @@ export const useQueryParams = <T extends RecordWithId>({
     queryParams,
   };
 };
+
+export interface QueryParamsStateNew<T extends RecordWithId> {
```
Comment (mark-prins) [inline packages/common/src/hooks/useQueryParams/useQueryParams.ts:71]: temporary, so I don't have to refactor everything in one big PR

File: packages/common/src/ui/layout/tables/context/TableContext.tsx [hunk 3]
```diff
@@ -30,8 +37,30 @@ export interface TableStore {
...
+const { Provider, useStore: useTableStore } = createContext<TableStore>();
+
+const TableProvider = <T extends RecordWithId>({
+  children,
+  queryParamsStore,
+  ...props
+}: PropsWithChildren<{
+  initialStore?: UseBoundStore<TableStore>;
+  createStore: () => UseBoundStore<TableStore>;
+  queryParamsStore?: UseBoundStore<QueryParamsStateNew<T>>;
```
Comment (mark-prins) [inline packages/common/src/ui/layout/tables/context/TableContext.tsx:49]: intending to make this required when all usages are updated

File: packages/invoices/src/OutboundShipment/ListView/ListView.tsx [hunk 4]
```diff
@@ -85,7 +79,7 @@ export const OutboundShipmentListViewComponent: FC = () => {
 
       <DataTable
         pagination={{ ...pagination, total: data?.totalCount }}
-        onChangePage={onChangePage}
+        onChangePage={pagination.onChangePage}
```
Comment (mark-prins) [inline packages/invoices/src/OutboundShipment/ListView/ListView.tsx:82]: changed the shape of the object a little too.. I haven't found a good reason to retain the nesting & duplication

**Review by clemens-msupply — APPROVED** [review]
looks good to me

**Pull request merged**